materials_io/utils/__init__.py

```diff
@@ -0,0 +1,178 @@
+            return None
+
+    # coerce empty values to None
+    if sub_dict in [{}, dict(), [], '', None]:
+        return None
+
+    if cast is not None:
+        return cast(sub_dict)
+    else:
+        return sub_dict
+
+
+def set_nested_dict_value(nest_dict: Dict, path: Tuple,
+                          value: Any, override: Optional[bool] = False,):
+    """
+    Set a value within a nested dictionary structure by traversing into
+    the dictionary as deep as that path found and changing it to ``value``.
+    If ``value`` is ``None``, immediately return without performing an action
+    Cribbed from https://stackoverflow.com/a/13688108/1435788
+
+    Parameters
+    ----------
+    nest_dict
+        A dictionary of dictionaries that is to be queried
+    path
+        A tuple (or other iterable type) that specifies the subsequent keys
+        needed to get to a value within `nest_dict`
+    value
+        The value which will be given to the path in the nested dictionary
+    override
+        If the value is already present, this flag controls whether to override
+        its existing value
+    """
+    if value is None:
+        return
+    orig_value = get_nested_dict_value_by_path(nest_dict, path)
+    for key in path[:-1]:
+        nest_dict = nest_dict.setdefault(key, {})
+    if orig_value is None or \
+            orig_value is not None and override:
+        # only set the value if it was None, or we chose to override
+        nest_dict[path[-1]] = value
+
+
+def set_nested_dict_value_with_units(nest_dict: Dict, path: Tuple,
+                                     value: Any, units: Optional[str] = None,
+                                     override: bool = False,
+                                     fn: Optional[Callable] = None):
+    """
+    Same as :func:`~materials_io.utils.set_nested_dict_value`, but sets the
+    value in the format of a dictionary with keys ``'value'`` and ``'units'``
+    according to the specified units. If ``fn`` is supplied, it will be
+    applied to the value prior to setting it.
+    """
+    if value is not None:
+        if fn is not None:
+            value = fn(value)
+        to_set = {'value': value}
+        if units is not None:
+            to_set['units'] = units
+        set_nested_dict_value(nest_dict, path, to_set, override)
+
+
+# type definition for the mapping dictionaries
+MappingElements = TypedDict('MappingElements',
+                            {'source_dict': Dict,
+                             'source_path': Union[str, Tuple[str, ...]],
+                             'dest_dict': Dict,
+                             'dest_path': Union[str, Tuple[str, ...]],
+                             'cast_fn': Optional[Callable],
+                             'units': Optional[Union[None, str]],
+                             'conv_fn': Optional[Union[None, Callable]],
+                             'override': Optional[bool]})
+
+
+def map_dict_values(mapping: List[MappingElements]):
```

Comment:
This is interesting way of approaching things. I don't yet fully appreciate it, but it feels like we'd be able to use this "dict to dict" mapping many places.

Reply:
When I started doing that, I was hoping it would result in less-verbose code overall, but then I added the `cast` and `conv` options and it got sort of out of control :sweat_smile: It was inspired by the approach used by HyperSpy